src/main/scala/format/pgn/Node.scala

```diff
@@ -0,0 +1,96 @@
+package chess
+package format
+package pgn
+
+import cats.syntax.all.*
+import cats.*
+
+import monocle.{ Optional, Traversal }
+
+// question: what if a variation without moves?
+// answer: it's possible but we need to support null move first
+case class Node[A](
+    value: A,
+    child: Option[Node[A]],   // main line next move
+    variations: List[Node[A]] // alternate moves, e.g. Node("e4", None, List("d4", "Nf3"))
+):
+  def mainLine: List[A] = value :: child.fold(Nil)(_.mainLine)
+  def totalNodes: Int   = this.foldLeft(0)((b, _) => b + 1)
+
+  // find the first Node that statisfies the predicate
+  def findNode(predicate: A => Boolean): Option[Node[A]] =
+    if predicate(value) then this.some
+    else
+      child
+        .flatMap(_.findNode(predicate))
+        .orElse(variations.foldLeft(none[Node[A]])((b, n) => b.orElse(n.findNode(predicate))))
+
+  def replaceNode(predicate: A => Boolean)(node: Node[A]): Option[Node[A]] =
+    modifyNode(predicate)(_ => node)
+
+  // modify the first node that satisfies the predicate (dfs with main line and then variations)
+  def modifyNode(predicate: A => Boolean)(f: Node[A] => Node[A]): Option[Node[A]] =
+    if predicate(value) then f(this).some
+    else
+      child.flatMap(_.modifyNode(predicate)(f)) match
+        case Some(n) => copy(child = Some(n)).some
+        case None =>
+          variations.foldLeft((false, List.empty[Node[A]])) {
+            case ((true, acc), n) => (true, acc :+ n)
+            case ((false, acc), n) =>
+              n.modifyNode(predicate)(f) match
+                case Some(nn) => (true, acc :+ nn)
+                case None     => (false, acc :+ n)
+          } match
+            case (true, ns) => copy(variations = ns).some
+            case (false, _) => none
+
+  // delete the first node that satisfies the predicate (both child and variations)
+  // except the root
+  def deleteSubNode(predicate: A => Boolean): Option[Node[A]] =
```

Comment:
This is a bit odds, compare to other functions, but I guess it's fine since deleting the root makes no sense